File: 10_Graphs/Traversal/Traversal.cpp
#include <iostream>
#include <vector>
#include <list>
#include <queue>

using namespace std;

class Graph {
    int V;
    list<int>* adj;

public:
    Graph(int v) : V(v) {
        adj = new list<int>[V];
    }

    void addEdge(int v, int w) {
        adj[v].push_back(w);
    }

    void BFS(int s) {
        vector<bool> visited(V, false);
        queue<int> q;

        visited[s] = true;
        q.push(s);

        cout << "BFS starting from vertex " << s << ": ";
        while (!q.empty()) {
            s = q.front();
            cout << s << " ";
            q.pop();

            for (auto adj_vertex : adj[s]) {
                if (!visited[adj_vertex]) {
                    visited[adj_vertex] = true;
                    q.push(adj_vertex);
                }
            }
        }
        cout << endl;
    }

    void DFSUtil(int v, vector<bool>& visited) {
        visited[v] = true;
        cout << v << " ";

        for (auto adj_vertex : adj[v])
            if (!visited[adj_vertex])
                DFSUtil(adj_vertex, visited);
    }

    void DFS(int v) {
        vector<bool> visited(V, false);
        cout << "DFS starting from vertex " << v << ": ";
        DFSUtil(v, visited);
        cout << endl;
    }
};

int main() {
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 0);
    g.addEdge(2, 3);
    g.addEdge(3, 3);

    g.BFS(2);
    g.DFS(2);

    return 0;
}

File: 10_Graphs/Traversal/Traversal.py
from collections import defaultdict, deque

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def bfs(self, s):
        visited = set()
        queue = deque([s])
        visited.add(s)

        print(f"BFS starting from vertex {s}: ", end="")
        while queue:
            vertex = queue.popleft()
            print(vertex, end=" ")

            for neighbor in self.graph[vertex]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        print()

    def dfs_util(self, v, visited):
        visited.add(v)
        print(v, end=" ")

        for neighbor in self.graph[v]:
            if neighbor not in visited:
                self.dfs_util(neighbor, visited)

    def dfs(self, v):
        visited = set()
        print(f"DFS starting from vertex {v}: ", end="")
        self.dfs_util(v, visited)
        print()

# Sample usage
g = Graph()
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(1, 2)
g.add_edge(2, 0)
g.add_edge(2, 3)
g.add_edge(3, 3)

g.bfs(2)
g.dfs(2)
